### DISCUSSION TITLE
[plugin] API Billing

### DISCUSSION THREAD

**Discussion by sonicaghi** [Kong/kong#7362] [Ideas and feature requests]
first version with just a Stripe connection.

**Comment by jusw85** [comment]
:+1:

**Comment by Tieske** [comment]
> 1) How do your customers typically integrate Gelato with billing?  I see that you support dev on boarding which is great to get people on and start playing with the API however our data providers charge us for accessing data and so for the customer to use real data they would need to purchase a credit package and be assigned an API key.  I am looking at Chargify as a possible billing solution, do you have anyone else integrating Gelato with Chargify or have any other reference cases I can use?

Billing plugin for Kong people asking for that in Gelato support cc @nikz

**Comment by mytototo** [comment]
Any news regarding an open-source billing plugin for Kong?

**Comment by TimDaub** [comment]
Just wanted to say that we researched using Kong for our API, but lack of payment integration is a show-stopper for us :/

**Comment by mytototo** [comment]
Agreed... Even if this can be considered as a portal feature and not really a gateway one, this is definitely something missing in Kong.

**Comment by javeme** [comment]
We need this feature.

**Comment by TimMcCauley** [comment]
How are Kong users dealing with the issue currently?

**Comment by dobrych** [comment]
For simple pricing cases, it's enough to set rate limits and analyze logs against specific usage patterns. It's possible to feed Kong's logs into ElasticSearch (for example) and do periodic processing to determine usage and trigger actions. Alternatively, it's possible to derive metrics from logs into some time-series DB too and trigger actions to produce invoices out of it.

Let's discuss what would be most common pricing schemes because billing is always about usage->$ formula.

**Comment by tumluliu** [comment]
thanks a lot for your hints! @dobrych

**Comment by TimMcCauley** [comment]
@dobrych the API we are currently working on will have different payment plans. As you already mentioned we are planning to restrict a user of a plan to a specfic ACL (most basic use case: _n_ requests for _x_ $). When the user signs up (probably going to use gelato) he/she will have to either pay a monthly one-time fee or subscribe to the plan. I was wondering which service/plugin Kong users would suggest to achieve this payment step? Is anyone consuming [chargify](https://www.chargify.com/pricing/) as an API?

**Comment by restyler** [comment]
What a nice thread which I've been tracking for 2 years already.. 
I am bootstrapping ApiRoad.net, which combines a dashboard and reverse proxy (API gateway). As an API vendor on RapidAPI myself, I always wanted to properly implement usage-based API billing and combine it with adequate logs and analytics. It was a challenging project in terms of development efforts and time spent, where I have combined ClickHouse for logs & billing; and Stripe for recurring payments.  I have been using ApiRoad for my own APIs for a few months now, and I am finally starting to roll out vendor access so that anyone can list their API, design plans, and get proper billing for it. I will be happy to chat with people who are seeking a solution for API billing, please drop me a message: contact@apiroad.net

**Comment by rezaloo75** [comment]
Thanks for being a Kong user @restyler 

API Billing, is not just an API Gateway concern and is in the realm of API productization. Core features needed for API Productization are brand-customizable developer portal, developer/app registration, and usage metering, all of which are available with Kong Gateway and via Kong Konnect - you can check it out [here](https://konghq.com/products/kong-konnect/register). On top of these requirements sometimes comes pricing & 3rd party billing (the subject of this feature request) although these can be implemented today in Kong via custom plugins, the request for OOTB integration is a fair one and we have plans to address it in Kong Konnect. You might also want to checkout [this video](https://konghq.com/resources/videos/transforming-telecommunications-using-kong-monetization) that shows how Telecom Indonesia used Kong Gateway for monetization and billing of their APIs.
